verilog/tools/kythe/indexing_facts_tree_extractor.cc

```diff
@@ -1105,20 +1113,37 @@ void IndexingFactsTreeExtractor::ExtractClassInstances(
   IndexingFactNode type_node(
       IndexingNodeData{IndexingFactType::kDataTypeReference});
 
-  const verible::SyntaxTreeLeaf* type =
+  const verible::Symbol* type =
       GetTypeIdentifierFromDataDeclaration(data_declaration_node);
   if (type == nullptr) {
     return;
   }
-  const Anchor type_anchor(type->get(), context_.base);
-  type_node.Value().AppendAnchor(type_anchor);
 
-  // Extract parameter list
-  const SyntaxTreeNode* param_list =
-      GetParamListFromDataDeclaration(data_declaration_node);
-  if (param_list != nullptr) {
-    const IndexingFactsTreeContext::AutoPop p(&facts_tree_context_, &type_node);
-    Visit(*param_list);
+  const auto tag = static_cast<NodeEnum>(type->Tag().tag);
+  if (tag == NodeEnum::kQualifiedId) {
+    // Then this is a qualified id node.
+    // e.g. "pkg::my_class var_name".
+    type->Accept(this);
+    MoveAndDeleteLastSibling(type_node);
+  } else if (tag == NodeEnum::kUnqualifiedId) {
+    // Then this is an unqualified id node.
+    // e.g "my_class var_name".
+    const SyntaxTreeLeaf& type_identifier =
+        *ABSL_DIE_IF_NULL(AutoUnwrapIdentifier(*type));
+    type_node.Value().AppendAnchor(
+        Anchor(type_identifier.get(), context_.base));
+
+    // Extract parameter list
+    const SyntaxTreeNode* param_list =
+        GetParamListFromDataDeclaration(data_declaration_node);
```

Comment:
Wouldn't it make more sense to get the ParamList directly from the unqualified id node at this stage, since you've already descended this far?

Also is this functionality duplicated elsewhere that handles arbitrary types (qualified, unqualified), basically anywhere there's a type reference (extends, type parameters, typedef, data declarations)?

Reply:
I refactored some parts but the remaining parts can't be simply refactored.